<a href="https://colab.research.google.com/github/BaishaliChetia/CapsNet-Keras/blob/master/latencyComparison.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Original implementation at:

https://github.com/ageron/handson-ml/blob/master/extra_capsnets.ipynb

Geron's model doesn't use the keras functional API. In the keras functional API, you don't need to give the batchsize. 

When you print the model, you get this:

```
Layer (type)                 Output Shape              Param #   
_________________________________________________________________
input (InputLayer)           [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv_layer_1 (Conv2D)        (None, 20, 20, 256)       20992     
_________________________________________________________________
conv_layer_2 (Conv2D)        (None, 6, 6, 256)         5308672   
_________________________________________________________________
reshape_layer_1 (Reshape)    (None, 1, 1152, 8)        0         
_________________________________________________________________
caps1_output_layer (SquashLa (None, 1, 1152, 8)        0         
_________________________________________________________________
Total params: 5,329,664
Trainable params: 5,329,664
Non-trainable params: 0
```

Notice that the Input-layer has shape (None, 28, 28, 1), but we only specified (28, 28, 1). It added None implicitly and that takes care of the batch.

So for anywhere Geron uses the batch size explicitly, you don't need to do anything and tensorflow will take care of.

Also note that tensorflow 1 APIs are still provided with the compat layer. I used the reduce_sum from TF1 in the squash layer, that allowed me to use Geron's code.

Documentation on how to migrate from TF1 to TF2 can be found here:

https://www.tensorflow.org/guide/migrate


In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0x1
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 b

In [ ]:
!df -h

Filesystem      Size  Used Avail Use% Mounted on
overlay         148G   38G  110G  26% /
tmpfs            64M     0   64M   0% /dev
tmpfs           6.4G     0  6.4G   0% /sys/fs/cgroup
shm             5.8G     0  5.8G   0% /dev/shm
/dev/sda1       154G   41G  114G  27% /opt/bin/.nvidia
tmpfs           6.4G   28K  6.4G   1% /var/colab
tmpfs           6.4G     0  6.4G   0% /proc/acpi
tmpfs           6.4G     0  6.4G   0% /proc/scsi
tmpfs           6.4G     0  6.4G   0% /sys/firmware


In [ ]:
!cat /proc/meminfo

MemTotal:       13302924 kB
MemFree:        10364448 kB
MemAvailable:   12461016 kB
Buffers:          146912 kB
Cached:          2089404 kB
SwapCached:            0 kB
Active:          1046864 kB
Inactive:        1598036 kB
Active(anon):     377796 kB
Inactive(anon):      480 kB
Active(file):     669068 kB
Inactive(file):  1597556 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:              4512 kB
Writeback:             0 kB
AnonPages:        408660 kB
Mapped:           239712 kB
Shmem:              1164 kB
KReclaimable:     142836 kB
Slab:             193356 kB
SReclaimable:     142836 kB
SUnreclaim:        50520 kB
KernelStack:        4816 kB
PageTables:         5528 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     6651460 kB
Committed_AS:    3016412 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       45552 kB
VmallocChunk:          0 kB
Percpu:          

In [ ]:
!nvidia-smi
print()
print()
print('=' * 80)
print()
print()
!df -h
print()
print()
print('=' * 80)
print()
print()
!free -m
print()
print()
print('=' * 80)
print()
print()
!lscpu
print()
print()
print('=' * 80)
print()
print()
! ps -eo pmem,pcpu,vsize,pid,cmd | sort -k 1 -nr | head -5

Thu Aug 19 20:46:51 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   38C    P0    27W / 250W |      0MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import numpy as np
import tensorflow as tf
import pandas as pd
import tensorflow.keras as K
#import tensorflow_model_optimization as tfmot
import keras

In [ ]:
pip install -q tensorflow-model-optimization

     |████████████████████████████████| 211 kB 8.7 MB/s 


In [ ]:
import tensorflow_model_optimization as tfmot

# Testing Latency for VGG16 

In [ ]:
tf.random.set_seed(500000)
# Load MNIST dataset
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the input image so that each pixel value is between 0 and 1.
train_images = train_images / 255.0
test_images = test_images / 255.0
train_X =[]

for i in range(train_images.shape[0]):
  train_pad = np.pad(train_images[i], pad_width=10 , mode= 'edge')
  #print(train_pad)
  train_X.append(train_pad)
#PADDING test images
test_X = []
for i in range(test_images.shape[0]):
  test_pad = np.pad(test_images[i], pad_width=10 , mode= 'edge')
  #print(train_pad)
  test_X.append(test_pad)
train_X = np.array(train_X)
test_X = np.array(test_X)


In [ ]:
import keras
tf.random.set_seed(500000)
# Load MNIST dataset
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = K.datasets.mnist.load_data()

# Normalize the input image so that each pixel value is between 0 and 1.
train_images = train_images / 255.0
test_images = test_images / 255.0
train_X =[]

for i in range(train_images.shape[0]):
  train_pad = np.pad(train_images[i], pad_width=2 , mode= 'edge')
  #print(train_pad)
  train_X.append(train_pad)
#PADDING test images
test_X = []
for i in range(test_images.shape[0]):
  test_pad = np.pad(test_images[i], pad_width=2 , mode= 'edge')
  #print(train_pad)
  test_X.append(test_pad)
train_X = np.array(train_X)
test_X = np.array(test_X)

In [ ]:
train_X.shape

(60000, 32, 32)

In [ ]:
train_X=np.dstack([train_X] * 3)
test_X=np.dstack([test_X]*3)
train_X.shape,test_X.shape

((60000, 32, 96), (10000, 32, 96))

In [ ]:
train_X = train_X.reshape(-1, 32,32,3)
test_X= test_X.reshape (-1,32,32,3)

train_X = train_X.astype('float32')
test_X = test_X.astype('float32')
train_X.shape,test_X.shape

((60000, 32, 32, 3), (10000, 32, 32, 3))

In [ ]:
vggModel = tf.keras.applications.VGG16(weights='imagenet', include_top=False, input_shape=(32, 32, 3))
vggModel.trainable = True
model2 = keras.Sequential([
  vggModel,
  keras.layers.Flatten(),
  tf.keras.layers.Dropout(0.5),
  keras.layers.Dense(256, activation='relu'),
  keras.layers.Dense(50, activation = 'relu'),
  keras.layers.Dense(10, activation = 'softmax')
])
# Train the digit classification model
model2.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# history = model2.fit(
#   train_X,
#   train_labels,
#   epochs= 50,
#   validation_split=0.1, callbacks = callback_list
# )


In [ ]:
train_X.shape

(60000, 48, 48, 3)

In [ ]:
import time
class TimeHistory(K.callbacks.Callback):
    def on_train_begin(self, logs={}):
        self.times = []

    def on_epoch_begin(self, batch, logs={}):
        self.epoch_time_start = time.time()

    def on_epoch_end(self, batch, logs={}):
        self.times.append(time.time() - self.epoch_time_start)
time_callback = TimeHistory()
history = model2.fit( train_X,train_labels, epochs= 20,validation_split=0.1, callbacks = [time_callback])
times = time_callback.times
print(times)

Epoch 1/20


/usr/local/lib/python3.7/dist-packages/keras/backend.py:4907: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a sigmoid or softmax activation and thus does not represent logits. Was this intended?"
  '"`sparse_categorical_crossentropy` received `from_logits=True`, but '


1688/1688 [==============================] - 34s 19ms/step - loss: 0.9943 - accuracy: 0.6343 - val_loss: 0.1315 - val_accuracy: 0.9708
Epoch 2/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.1644 - accuracy: 0.9611 - val_loss: 0.1204 - val_accuracy: 0.9762
Epoch 3/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.1414 - accuracy: 0.9679 - val_loss: 0.0710 - val_accuracy: 0.9873
Epoch 4/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.1026 - accuracy: 0.9778 - val_loss: 0.0763 - val_accuracy: 0.9810
Epoch 5/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.0995 - accuracy: 0.9795 - val_loss: 0.0745 - val_accuracy: 0.9837
Epoch 6/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.0729 - accuracy: 0.9846 - val_loss: 0.1108 - val_accuracy: 0.9872
Epoch 7/20
1688/1688 [==============================] - 32s 19ms/step - loss: 0.1275 - accuracy: 0.9769 - val_loss: 0.0911 - val_accura

In [ ]:
time_per_epoch = sum(times)/len(times)
print(time_per_epoch)

32.194143211841585


# For CNN(VGG16), training time per epoch is 32.19 seconds
## 0.59ms per training instance


In [ ]:
test_X.shape

(10000, 32, 32, 3)

In [ ]:
_, baseline_model_accuracy = model2.evaluate(
    test_X, test_labels,batch_size = 32, verbose=1, callbacks = [time_callback] )


313/313 [==============================] - 2s 7ms/step - loss: 0.0403 - accuracy: 0.9927


# Test time = 2ms per test instance

In [ ]:
keras_file = "/content/drive/MyDrive/MnistResults/vgg32X32basemodel.h5"
tf.keras.models.save_model(model2, keras_file, include_optimizer=False)
print('Saved baseline model to:', keras_file)

Saved baseline model to: /content/drive/MyDrive/MnistResults/vgg32X32basemodel.h5


In [ ]:
import tensorflow_model_optimization as tfmot
import tempfile

prune_low_magnitude = tfmot.sparsity.keras.prune_low_magnitude

# TODO: Remove this line
#m.load_weights('/content/drive/MyDrive/MnistResults/best_weights4.hdf5')

# print("ORIGINAL MODEL")
# print(mm.summary())
# print('-' * 80)

# Compute end step to finish pruning after 2 epochs.
batch_size = 128
epochs = 2
validation_split = 0.1 # 10% of training set will be used for validation set. 

num_images = x_train.shape[0] * (1 - validation_split)
end_step = np.ceil(num_images / batch_size).astype(np.int32) * epochs

model2.load_weights('/content/drive/MyDrive/MnistResults/vggtrain_best_weights2.hdf5')
pruning_params = {
      'pruning_schedule': tfmot.sparsity.keras.PolynomialDecay(initial_sparsity=0.50,
                                                              final_sparsity=0.80,
                                                              begin_step=0,
                                                              end_step=end_step)
}

# Define model for pruning.
model_for_pruning = prune_low_magnitude(model2, **pruning_params)

# `prune_low_magnitude` requires a recompile.
model_for_pruning.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

#model_for_pruning.summary()
logdir = tempfile.mkdtemp()

callbacks = [
  tfmot.sparsity.keras.UpdatePruningStep(),
  tfmot.sparsity.keras.PruningSummaries(log_dir=logdir),
]

model_for_pruning.fit(train_X, train_labels,
                  batch_size=batch_size, epochs=epochs, validation_split=validation_split,
                  callbacks=callbacks)
_, model_for_pruning_accuracy = model_for_pruning.evaluate(
  test_X, test_labels, verbose=0)

/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py:2223: UserWarning: `layer.add_variable` is deprecated and will be removed in a future version. Please use `layer.add_weight` method instead.
  warnings.warn('`layer.add_variable` is deprecated and '


Epoch 1/2


/usr/local/lib/python3.7/dist-packages/keras/backend.py:4907: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a sigmoid or softmax activation and thus does not represent logits. Was this intended?"
  '"`sparse_categorical_crossentropy` received `from_logits=True`, but '


422/422 [==============================] - 26s 48ms/step - loss: 0.9248 - accuracy: 0.6368 - val_loss: 0.1801 - val_accuracy: 0.9543
Epoch 2/2
422/422 [==============================] - 18s 44ms/step - loss: 0.0922 - accuracy: 0.9796 - val_loss: 0.0441 - val_accuracy: 0.9908


In [ ]:
test_X.shape

(10000, 48, 48, 3)

In [ ]:
_, model_for_pruning_accuracy = model_for_pruning.evaluate(
  test_X, test_labels, verbose=1)

313/313 [==============================] - 2s 8ms/step - loss: 0.0581 - accuracy: 0.9887


In [ ]:
model_for_export = tfmot.sparsity.keras.strip_pruning(model_for_pruning)

_, pruned_keras_file = tempfile.mkstemp('.h5')
tf.keras.models.save_model(model_for_export, pruned_keras_file, include_optimizer=False)
print('Saved pruned Keras model to:', pruned_keras_file)

Saved pruned Keras model to: /tmp/tmp968lwznm.h5


In [ ]:
model_for_export.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
_, model_for_export.evaluate(test_X, test_labels, verbose=1)

/usr/local/lib/python3.7/dist-packages/keras/backend.py:4907: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a sigmoid or softmax activation and thus does not represent logits. Was this intended?"
  '"`sparse_categorical_crossentropy` received `from_logits=True`, but '


313/313 [==============================] - 3s 7ms/step - loss: 0.0581 - accuracy: 0.9887


(56, [0.05805013328790665, 0.9886999726295471])

In [ ]:
_, model_for_export.evaluate(test_X, test_labels, verbose=1)

313/313 [==============================] - 2s 7ms/step - loss: 0.0581 - accuracy: 0.9887


(((56, [0.05805013328790665, 0.9886999726295471]),
  [0.05805013328790665, 0.9886999726295471]),
 [0.05805013328790665, 0.9886999726295471])

In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model_for_export)
pruned_tflite_model = converter.convert()

_, pruned_tflite_file = tempfile.mkstemp('.tflite')

with open(pruned_tflite_file, 'wb') as f:
  f.write(pruned_tflite_model)

print('Saved pruned TFLite model to:', pruned_tflite_file)

INFO:tensorflow:Assets written to: /tmp/tmp79yckqty/assets
Saved pruned TFLite model to: /tmp/tmpxntgk577.tflite


# get size of zipped quantized and pruned model

In [ ]:
def get_gzipped_model_size(file):
  # Returns size of gzipped model, in bytes.
  import os
  import zipfile

  _, zipped_file = tempfile.mkstemp('.zip')
  with zipfile.ZipFile(zipped_file, 'w', compression=zipfile.ZIP_DEFLATED) as f:
    f.write(file)

  return os.path.getsize(zipped_file)

In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model_for_export)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
quantized_and_pruned_tflite_model = converter.convert()

_, quantized_and_pruned_tflite_file = tempfile.mkstemp('.tflite')

with open(quantized_and_pruned_tflite_file, 'wb') as f:
  f.write(quantized_and_pruned_tflite_model)

print('Saved quantized and pruned TFLite model to:', quantized_and_pruned_tflite_file)

print("Size of gzipped baseline Keras model: %.2f bytes" % (get_gzipped_model_size(keras_file)))
print("Size of gzipped pruned and quantized TFlite model: %.2f bytes" % (get_gzipped_model_size(quantized_and_pruned_tflite_file)))

INFO:tensorflow:Assets written to: /tmp/tmpz09n_pe1/assets


INFO:tensorflow:Assets written to: /tmp/tmpz09n_pe1/assets


Saved quantized and pruned TFLite model to: /tmp/tmpdgc0ta08.tflite
Size of gzipped baseline Keras model: 17463391.00 bytes
Size of gzipped pruned and quantized TFlite model: 4851103.00 bytes


# Get Size of Pruned and Quantized model

In [ ]:
def get_tflite_model_size(model):
  # Returns size of gzipped model, in bytes.
  import os
  #import zipfile

  _, file_path = tempfile.mkstemp('.tflite')
  with open(file_path, 'wb') as f:
    f.write(model)

  return os.path.getsize(file_path)

In [ ]:
print("Size of gzipped pruned and quantized TFlite model: %.2f bytes" % (get_tflite_model_size(quantized_and_pruned_tflite_model)))

Size of gzipped pruned and quantized TFlite model: 14970816.00 bytes


In [ ]:
def get_base_model_size(model):
  # Returns size of gzipped model, in bytes.
  import os
  #import zipfile

  _, file_path = tempfile.mkstemp('.tf')
  with open(file_path, 'wb') as f:
    f.write(model)

  return os.path.getsize(file_path)

In [ ]:
print("Size of base model: %.2f bytes" % (get_base_model_size(model2)))

TypeError: ignored

In [ ]:
quantized_and_pruned_tflite_model

In [ ]:
import numpy as np

def evaluate_model(interpreter):
  input_index = interpreter.get_input_details()[0]["index"]
  output_index = interpreter.get_output_details()[0]["index"]

  # Run predictions on ever y image in the "test" dataset.
  prediction_digits = []
  for i, test_image in enumerate(test_X):
    if i % 1000 == 0:
      print('Evaluated on {n} results so far.'.format(n=i))
    # Pre-processing: add batch dimension and convert to float32 to match with
    # the model's input data format.
    test_image = np.expand_dims(test_image, axis=0).astype(np.float32)
    interpreter.set_tensor(input_index, test_image)

    # Run inference.
    interpreter.invoke()

    # Post-processing: remove batch dimension and find the digit with highest
    # probability.
    output = interpreter.tensor(output_index)
    digit = np.argmax(output()[0])
    prediction_digits.append(digit)

  print('\n')
  # Compare prediction results with ground truth labels to calculate accuracy.
  prediction_digits = np.array(prediction_digits)
  accuracy = (prediction_digits == test_labels).mean()
  return accuracy

In [ ]:
interpreter = tf.lite.Interpreter(model_content=quantized_and_pruned_tflite_model)
interpreter.allocate_tensors()

test_accuracy = evaluate_model(interpreter)

print('Pruned and quantized TFLite test_accuracy:', test_accuracy)
print('Pruned TF test accuracy:', model_for_pruning_accuracy)

Evaluated on 0 results so far.


KeyboardInterrupt: ignored

# ====================================================================================

# Testing training latency for Capsule Network 

In [ ]:
tf.random.set_seed(500000)
# Load MNIST dataset
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the input image so that each pixel value is between 0 and 1.
train_images = train_images / 255.0
test_images = test_images / 255.0
train_X =[]

for i in range(train_images.shape[0]):
  train_pad = np.pad(train_images[i], pad_width=10 , mode= 'edge')
  #print(train_pad)
  train_X.append(train_pad)
#PADDING test images
test_X = []
for i in range(test_images.shape[0]):
  test_pad = np.pad(test_images[i], pad_width=10 , mode= 'edge')
  #print(train_pad)
  test_X.append(test_pad)
train_X = np.array(train_X)
test_X = np.array(test_X)


In [ ]:
train_X.shape, test_X.shape

((60000, 48, 48), (10000, 48, 48))

In [ ]:
x_train = train_X
x_test = test_X
y_train = train_labels
y_test = test_labels

In [ ]:
caps1_n_maps = 32
caps1_n_caps = caps1_n_maps * 16 * 16  # 1152 primary capsules
caps1_n_dims = 8
caps2_n_caps = 10
caps2_n_dims = 16

tf.random.set_seed(500000)

In [ ]:
#class SquashLayer(K.layers.Layer, tfmot.sparsity.keras.PrunableLayer):
class SquashLayer(K.layers.Layer):
  def __init__(self, axis=-1, **kwargs):
    super(SquashLayer, self).__init__(**kwargs)
    self.axis = axis
    
  def build(self, input_shapes):
    pass

  """
  def get_prunable_weights(self):
    return []
  """

  def call(self, inputs):
    EPSILON = 1.0e-9
    squared_norm = tf.compat.v1.reduce_sum(tf.square(inputs),\
                                           axis=self.axis,\
                                           keepdims=True)
    safe_norm = tf.sqrt(squared_norm + EPSILON)
    squash_factor = squared_norm / (1. + squared_norm)
    unit_vector = inputs / safe_norm
    return squash_factor * unit_vector

  def get_config(self):
    config = super(SquashLayer, self).get_config()
    config.update({"axis": self.axis})
    return config

 
#class SafeNorm(K.layers.Layer, tfmot.sparsity.keras.PrunableLayer):
class SafeNorm(K.layers.Layer):
  
  def __init__(self, axis=-1, keep_dims = False,  **kwargs):
    super(SafeNorm, self).__init__(**kwargs)
    self.axis = axis
    self.keep_dims = keep_dims

  def build(self, input_shapes):
    pass

  """
  def get_prunable_weights(self):
    return []
  """

  def call(self, input):
    EPSILON = 1.0e-9
    squared_norm = tf.compat.v1.reduce_sum(tf.square(inputs),\
                                           axis=self.axis,\
                                           keepdims= self.keep_dims)
    safe_norm = tf.sqrt(squared_norm + EPSILON)
    return safe_norm

  def get_config(self):
    config = super(SafeNorm, self).get_config()
    config.update({"axis": self.axis, "keep_dims": self.keep_dims})
    return config
  
# This should be the part where the digit layer, and where we tile things
# This is incomplete, and work in progress
# TODO: Complete this
class MyDigitCapsLayer(K.layers.Layer, tfmot.sparsity.keras.PrunableLayer):
  def __init__(self, **kwargs):
    super(MyDigitCapsLayer, self).__init__(**kwargs)
  
  def get_config(self):
    config =  super(MyDigitCapsLayer, self).get_config()
    return config
  

  def build(self, input_shapes):
    init_sigma = 0.1  # TODO: use
    self.kernel = self.add_weight(\
                      "kernel",\
                      (caps1_n_caps, caps2_n_caps, caps2_n_dims, caps1_n_dims),\
                      initializer="random_normal",\
                      dtype=tf.float32)

  
  # To debug this function, I used prints to print the shape
  # expand_dims just adds an exis, so if you say expand_dims(inshape=(5, 3), -1),
  # you get the output shape (5, 3, 1), it just adds an axis at the end
  # Then tile just multiplies one of the dimensions (that is it stacks along that direction N times)
  # so tile(inshape=(5, 3, 1), [1, 1, 1000]) will yield a shape (5, 3, 1000)
  #
  # Notice I didn't tile in build, but in call, Most probaly this is the right thing to do
  # but we'll only figure out when we actually train
  def get_prunable_weights(self):
    return [self.kernel]
  
  def call(self, inputs):
    # Add a dimension at the end
    exp1 = tf.expand_dims(inputs, -1, name="caps1_output_expanded")
    # add a dimension along 3rd axis
    exp1 = tf.expand_dims(exp1, 2, name="caps2_output_espanced")
    # tile along 3rd axis
    tile = tf.tile(exp1, [1, 1, caps2_n_caps, 1, 1], name="caps1_output_tiled")
    caps2_predicted = tf.matmul(self.kernel, tile, name="caps2_predicted")
    return caps2_predicted

# https://www.tensorflow.org/api_docs/python/tf/keras/losses/Loss
class MarginLoss(K.losses.Loss):
    def __init__(self, **kwargs):
      super(MarginLoss, self).__init__(**kwargs)

  
    def get_config(self):
      config = super(MarginLoss, self).get_config()
      return config
    
    def safe_norm(self, input, axis=-2, epsilon=1e-5, keep_dims=False, name=None):
      squared_norm = tf.reduce_sum(tf.square(input), axis=axis,
                                     keepdims=keep_dims)
      return tf.sqrt(squared_norm + epsilon)

    """
    def get_prunable_weights(self):
      return []
    """

    def call(self,y_true, input):
      # print(f"y_true.shape = {y_true.shape}, y_pred.shape = {y_pred.shape}")
      # return K.losses.MeanSquaredError()(y_true, y_pred)

      #y_true = K.Input(shape=[], dtype=tf.int64, name="y")
      m_plus = 0.9
      m_minus = 0.1
      lambda_ = 0.5 
      
      #y_true one hot encode y_train
      T = tf.one_hot(y_true, depth=caps2_n_caps, name="T")
      
      caps2_output_norm = self.safe_norm(input, keep_dims = True)

      present_error_raw = tf.square(\
                                    tf.maximum(0., m_plus - caps2_output_norm),
                                    name="present_error_raw")
      present_error = tf.reshape(\
                                    present_error_raw, shape=(-1, 10),
                                    name="present_error")  
  
      absent_error_raw = tf.square(\
                                    tf.maximum(0., caps2_output_norm - m_minus),
                                    name="absent_error_raw")
      absent_error = tf.reshape(\
                                    absent_error_raw, shape=(-1, 10),
                                    name="absent_error")
  
      L = tf.add(\
                  T * present_error,\
                  lambda_ * (1.0 - T) * absent_error,
                  name="L")
      marginLoss = tf.reduce_mean(\
                                  tf.reduce_sum(L, axis=1),\
                                  name="margin_loss")
      return marginLoss


#class RoutingByAgreement(K.layers.Layer, tfmot.sparsity.keras.PrunableLayer):
class RoutingByAgreement(K.layers.Layer):
  def __init__(self, round_number=-1, **kwargs):
    super(RoutingByAgreement, self).__init__(**kwargs)
    self.round_number = round_number 

  def get_config(self):
    config = super(RoutingByAgreement, self).get_config()
    config.update({"round_number": self.round_number})
    return config
 

  def build(self, input_shapes):
    self.raw_weights_1 = self.add_weight("raw_weights", \
                                         (caps1_n_caps, caps2_n_caps, 1, 1), \
                                         initializer = "zeros", \
                                         dtype=tf.float32,)
    
    #print("Routing layer: self.raw_weights = ", self.raw_weights.shape, "input_shape = ", input_shapes)

  
  def get_prunable_weights(self):
    return [self.raw_weights_1]


  @staticmethod
  def squash(inputs):
    EPSILON = 1.0e-5
    squared_norm = tf.compat.v1.reduce_sum(tf.square(inputs),\
                                           keepdims=True)
    safe_norm = tf.sqrt(squared_norm + EPSILON)
    squash_factor = squared_norm / (1. + squared_norm)
    unit_vector = inputs / safe_norm
    return squash_factor * unit_vector

  def single_round_routing(self, inputs, raw_weights, agreement):
    raw_weights = tf.add(raw_weights, agreement)
    routing_wt = tf.nn.softmax(raw_weights, axis=2)
    wt_predictions = tf.multiply(routing_wt, inputs)
    wt_sum = tf.reduce_sum(wt_predictions, axis=1, keepdims=True)
    return wt_sum

  def call(self, inputs):
    agreement = tf.zeros(shape=self.raw_weights_1.shape)
    sqsh_wt_sum = None
    x = inputs
    for i in range(2):
      wt_sum = self.single_round_routing(inputs, self.raw_weights_1, agreement)
      sqsh_wt_sum = RoutingByAgreement.squash(wt_sum)
      sqsh_wt_sum_tiled = tf.tile(\
                          sqsh_wt_sum ,\
                          [1, caps1_n_caps, 1, 1, 1],\
                          name="caps2_output_round_1_tiled")
      agreement = tf.matmul(\
                            x, \
                            sqsh_wt_sum_tiled,\
                            transpose_a=True,\
                            name="agreement")
    return sqsh_wt_sum

class MyAccuracy(K.metrics.Metric):
  def __init__(self, **kwargs):
    super(MyAccuracy, self).__init__(**kwargs)
    self.acc_obj = K.metrics.Accuracy()
    self.state = 0
  

  def get_config(self):
    config = super(MyAccuracy, self).get_config()
    config.update({"acc_obj": None, "state": self.state})
    return config
  

  def safe_norm(self, input, axis=-2, epsilon=1e-5, keep_dims=True, name=None):
      squared_norm = tf.reduce_sum(tf.square(input), axis=axis,
                                     keepdims=keep_dims)
      return tf.sqrt(squared_norm + epsilon)

  def update_state(self, y_true, input, sample_weight=None):
    if self.acc_obj is None:
      self.acc_obj = K.metrics.Accuracy()
    y_proba = self.safe_norm(input, axis=-2)
    y_proba_argmax = tf.argmax(y_proba, axis=2)
    y_pred = tf.squeeze(y_proba_argmax, axis=[1,2], name="y_pred")
    #y_true = tf.reshape(y_true, (y_true.shape[0], ))
    y_true = tf.cast(y_true, dtype=tf.int64)
    self.acc_obj.update_state(y_true, y_pred, sample_weight)

  def reset_state(self):
    self.acc_obj.reset_state()

  def result(self):
    return self.acc_obj.result()

class MyReshapeLayer(K.layers.Layer):
  def __init__(self, axis=-1, keep_dims = False,  **kwargs):
    super(MyReshapeLayer, self).__init__(**kwargs)

  def build(self, input_shapes):
    pass

  def safe_norm(self, input, axis=-2, epsilon=1e-5, keep_dims=True, name=None):
      squared_norm = tf.reduce_sum(tf.square(input), axis=axis,
                                     keepdims=keep_dims)
      return tf.sqrt(squared_norm + epsilon)

  def call(self, input):
    print('printing shapes ------------------- ')
    EPSILON = 1.0e-9
    print(input)
    y_proba = self.safe_norm(input, axis=-2)
    print(y_proba)
    y_proba_argmax = tf.argmax(y_proba, axis=2)
    print(y_proba_argmax)
    y_pred = tf.squeeze(y_proba_argmax, axis=[1,2], name="y_pred")
    print(y_pred)
    return tf.cast(y_pred, tf.int64)

  def get_config(self):
    config = super(MyReshapeLayer, self).get_config()
    return config
  

In [ ]:
class Model:
    @staticmethod
    def build(inshape=(48, 48, 1)):
        inp = K.Input(shape=inshape, dtype=tf.float32, name='input')
        
        # Primary capsules
        # For each digit in the batch
        # 32 maps, each 6x6 grid of 8 dimensional vectors
        
        # First Conv layer
        conv1_params = \
        {
            "filters": 256,
            "kernel_size": 9,
            "strides": 1,
            "padding": "valid",
            "activation": tf.nn.relu,
        }
        x = K.layers.Conv2D(**conv1_params, name="conv_layer_1")(inp)
        
        # Second conv layer
        conv2_params = \
        {
            "filters": caps1_n_maps * caps1_n_dims, # 256 convolutional filters
            "kernel_size": 9,
            "strides": 2,
            "padding": "valid",
            "activation": tf.nn.relu
        }
        x = K.layers.Conv2D(**conv2_params, name="conv_layer_2")(x)
        
        # Reshape
        x = K.layers.Reshape(\
                             (caps1_n_caps, caps1_n_dims),\
                             name="reshape_layer_1")(x)
                             
        x = SquashLayer(name="caps1_output_layer")(x)
        
        x = MyDigitCapsLayer(name = "caps2_predicted")(x)
        caps2_predicted = x # Save this value for later
        
        #routing by agreement (2 rounds)
        x = RoutingByAgreement(name="routing1", round_number=2)(x)
        
        return K.Model(inputs=inp, outputs=x, name='my')
    
m = Model.build()
print(m.summary())


Model: "my"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input (InputLayer)           [(None, 48, 48, 1)]       0         
_________________________________________________________________
conv_layer_1 (Conv2D)        (None, 40, 40, 256)       20992     
_________________________________________________________________
conv_layer_2 (Conv2D)        (None, 16, 16, 256)       5308672   
_________________________________________________________________
reshape_layer_1 (Reshape)    (None, 8192, 8)           0         
_________________________________________________________________
caps1_output_layer (SquashLa (None, 8192, 8)           0         
_________________________________________________________________
caps2_predicted (MyDigitCaps (None, 8192, 10, 16, 1)   10485760  
_________________________________________________________________
routing1 (RoutingByAgreement (None, 1, 10, 16, 1)      81920    

In [ ]:
m.compile(optimizer='adam', loss=MarginLoss(), metrics=[MyAccuracy()])

In [ ]:
history = m.fit(x_train, y_train, batch_size=32, epochs=2, verbose= 1, validation_split=0.1)

Epoch 1/2


TypeError: ignored

In [ ]:
class TimeHistory(K.callbacks.Callback):
    def on_train_begin(self, logs={}):
        self.times = []

    def on_epoch_begin(self, batch, logs={}):
        self.epoch_time_start = time.time()

    def on_epoch_end(self, batch, logs={}):
        self.times.append(time.time() - self.epoch_time_start)
time_callback = TimeHistory()
history = m.fit(x_train, y_train, batch_size=32, epochs=2, verbose= 1, validation_split=0.2, callbacks = [time_callback])
times = time_callback.times
print(times)

Epoch 1/2


ValueError: ignored

In [ ]:
print(sum(times)/len(times))


122.46780831813813


# Capsule Network training time per epoch is 122.46 seconds
# 2.5ms per training instance

# =========================================================================

In [ ]:
m.compile(optimizer='adam', loss=MarginLoss(), metrics=[MyAccuracy()])

In [ ]:
#Extra layer for evaluate
class DimensionCorrection(K.layers.Layer):
    def __init__(self, **kwargs):
      super(DimensionCorrection, self).__init__(**kwargs)
    
    def safe_norm(self, input, axis=-2, epsilon=1e-7, keep_dims=False, name=None):
      squared_norm = tf.reduce_sum(tf.square(input), axis=axis,
                                     keepdims=keep_dims)
      return tf.sqrt(squared_norm + epsilon)

    def call(self,y_pred):
      y_proba = self.safe_norm(y_pred, axis=-2)
      y_proba_argmax = tf.argmax(y_proba, axis=2)
      y_pred = tf.squeeze(y_proba_argmax, axis=[1,2], name="y_pred")
      return y_pred


In [ ]:
m.load_weights("/content/drive/MyDrive/MnistResults/caps_best_weightsA.hdf5")

In [ ]:
#m = Model.build()
m.load_weights('/content/drive/MyDrive/MnistResults/caps_best_weightsA.hdf5')
m.compile(optimizer='Adam', loss=MarginLoss)
newmodel = K.models.Sequential(\
            [\
            m,\
            DimensionCorrection(),\
            ]\
)
newmodel.summary()
m.trainable = False
newmodel.compile(optimizer='adam')

y_pred = newmodel.predict(x_test)

import sklearn
from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score
print(confusion_matrix(y_test, y_pred))
print(f"accuracy = {accuracy_score(y_test, y_pred)}")

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
my (Functional)              (None, 1, 10, 16, 1)      6815744   
_________________________________________________________________
dimension_correction (Dimens (None,)                   0         
Total params: 6,815,744
Trainable params: 6,815,744
Non-trainable params: 0
_________________________________________________________________
[[ 951   12    1    1    4    0    4    6    1    0]
 [   0 1131    1    2    0    0    0    1    0    0]
 [   0    4 1012    2    0    0    0   13    1    0]
 [   0    0    1 1002    0    1    0    5    1    0]
 [   0    4    0    0  973    0    4    0    0    1]
 [   0   11    1   24    1  839    5    6    5    0]
 [   0    7    0    0    1    1  946    0    3    0]
 [   0    4    6    1    0    0    0 1016    1    0]
 [   0    2    1    2    2    0    0    4  962    1]
 [   1   14    0   10   10   

In [ ]:
_, baseline_model_accuracy = m.evaluate(x_test, y_test, verbose = 1)

TypeError: ignored

In [ ]:
x_test.shape, y_test.shape

((10000, 28, 28), (10000,))

In [ ]:
import time

In [ ]:
x_test.shape[0]

10000

In [ ]:
y_pred = newmodel.predict(x_test)
print(y_pred)

[7 2 1 ... 4 5 6]


In [ ]:
t1 = time.perf_counter()
for i in range(x_test.shape[0]):
  y_pred[i] = newmodel.predict(x_test[i])
  print(f"accuracy = {accuracy_score(y_test[i], y_pred[i])}")
t2 = time.perf_counter()
per_instance_time = (t2-t1)/x_test.shape[0]
print(per_instance_time)


ValueError: ignored

In [ ]:
# pip install -q tensorflow-model-optimization

In [ ]:
comparison_metric.name


'my_accuracy_8'

In [ ]:
import tensorflow_model_optimization as tfmot
import tempfile

prune_low_magnitude = tfmot.sparsity.keras.prune_low_magnitude

# TODO: Remove this line
#m.load_weights('/content/drive/MyDrive/MnistResults/best_weights4.hdf5')

# print("ORIGINAL MODEL")
# print(mm.summary())
# print('-' * 80)

# Compute end step to finish pruning after 2 epochs.
batch_size = 128
epochs = 2
validation_split = 0.1 # 10% of training set will be used for validation set. 

num_images = x_train.shape[0] * (1 - validation_split)
end_step = np.ceil(num_images / batch_size).astype(np.int32) * epochs

mm = m
mm.load_weights('/content/drive/MyDrive/MnistResults/caps_best_weightsA.hdf5')

# Define model for pruning.
# pruning_params = {
#       'pruning_schedule': tfmot.sparsity.keras.PolynomialDecay(initial_sparsity=0.50,
#                                                                final_sparsity=0.80,
#                                                                begin_step=0,
#                                                                end_step=end_step)
# }

pruning_params = {
      'pruning_schedule': tfmot.sparsity.keras.PolynomialDecay(initial_sparsity=0.20,
                                                              final_sparsity=0.65,
                                                              begin_step=0,
                                                              end_step=end_step)
}

"""
model_for_pruning = prune_low_magnitude(mm, **pruning_params)

# `prune_low_magnitude` requires a recompile.
model_for_pruning.compile(optimizer='adam',
              loss=MarginLoss(), metrics=[MyAccuracy()])

model_for_pruning.summary()

"""


######################################################################
logdir = tempfile.mkdtemp()

callbacks = [
  tfmot.sparsity.keras.UpdatePruningStep(),
  tfmot.sparsity.keras.PruningSummaries(log_dir=logdir),
]

# Helper function uses `prune_low_magnitude` to make only the 
# Dense layers train with pruning.
def apply_pruning_to_layers(layer):
  #print("called")
  if isinstance(layer, MyDigitCapsLayer):
    print(f"Layer {layer} {layer.name} slated for pruning")
    return tfmot.sparsity.keras.prune_low_magnitude(layer, **pruning_params)
  elif layer.name == "conv_layer_2":
    print(f"Layer {layer} {layer.name} slated for pruning")
    return tfmot.sparsity.keras.prune_low_magnitude(layer, **pruning_params)
  elif layer.name == "routing1":
    print(f"Layer {layer} {layer.name} slated for pruning")
    return tfmot.sparsity.keras.prune_low_magnitude(layer, **pruning_params)
  elif layer.name == "conv_layer_1":
    print(f"Layer {layer} {layer.name} slated for pruning")
    return tfmot.sparsity.keras.prune_low_magnitude(layer, **pruning_params)


  print(f"Layer {layer} {layer.name} unchanged")
  return layer

# Use `tf.keras.models.clone_model` to apply `apply_pruning_to_layers` 
# to the layers of the model.
model_for_pruning = tf.keras.models.clone_model(
    mm,
    clone_function=apply_pruning_to_layers,
)
#print(model_for_pruning.summary())

"""
model_for_pruning = K.models.Sequential(\
              [\
                model_for_pruning,\
                MyReshapeLayer(),\
              ]\
)
"""


model_for_pruning.compile(optimizer='adam', loss=MarginLoss(), metrics=[MyAccuracy()])
#model_for_pruning.compile(optimizer='adam', loss=MarginLoss())

model_for_pruning.fit(x_train, y_train,
                  batch_size=32, epochs=2, validation_split=validation_split,
                  callbacks=callbacks)
_, model_for_pruning_accuracy = model_for_pruning.evaluate(x_test, y_test, verbose=1)


Layer <keras.layers.convolutional.Conv2D object at 0x7ff676053810> conv_layer_1 slated for pruning
Layer <keras.layers.convolutional.Conv2D object at 0x7ff676053c10> conv_layer_2 slated for pruning
Layer <keras.layers.core.Reshape object at 0x7ff676048390> reshape_layer_1 unchanged
Layer <__main__.SquashLayer object at 0x7ff676028a10> caps1_output_layer unchanged
Layer <__main__.MyDigitCapsLayer object at 0x7ff67616ef10> caps2_predicted slated for pruning
Layer <__main__.RoutingByAgreement object at 0x7ff5e67eba10> routing1 slated for pruning


/usr/local/lib/python3.7/dist-packages/keras/engine/base_layer.py:2223: UserWarning: `layer.add_variable` is deprecated and will be removed in a future version. Please use `layer.add_weight` method instead.
  warnings.warn('`layer.add_variable` is deprecated and '


Epoch 1/2
1688/1688 [==============================] - 71s 41ms/step - loss: 2.2779 - my_accuracy_1: 0.9723 - val_loss: 2.2753 - val_my_accuracy_1: 0.9722
Epoch 2/2
313/313 [==============================] - 10s 33ms/step - loss: 2.2767 - my_accuracy_1: 0.9733


In [ ]:
x_test.shape, y_test.shape

((10000, 28, 28), (10000,))

In [ ]:
_, model_for_pruning_accuracy = model_for_pruning.evaluate(x_test, y_test, verbose=1)

313/313 [==============================] - 10s 33ms/step - loss: 2.2767 - my_accuracy_1: 0.9733



#explanation for increase in parameters after pruning

https://www.dlology.com/blog/how-to-compress-your-keras-model-x5-smaller-with-tensorflow-model-optimization/

In [ ]:
# _, model_for_pruning_accuracy = model_for_pruning.evaluate(
#    x_test, y_test, verbose=1)

313/313 [==============================] - 11s 34ms/step - loss: 2.2769 - my_accuracy_72: 0.5371


In [ ]:
model_for_export = tfmot.sparsity.keras.strip_pruning(model_for_pruning)
model_for_export.summary()
pruned_keras_file = "/content/drive/MyDrive/MnistResults/caps_pruned_fileA.h5"
tf.keras.models.save_model(model_for_export, pruned_keras_file, include_optimizer=False)
print('Saved pruned Keras model to:', pruned_keras_file)

Model: "my"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input (InputLayer)           [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv_layer_1 (Conv2D)        (None, 20, 20, 256)       20992     
_________________________________________________________________
conv_layer_2 (Conv2D)        (None, 6, 6, 256)         5308672   
_________________________________________________________________
reshape_layer_1 (Reshape)    (None, 1152, 8)           0         
_________________________________________________________________
caps1_output_layer (SquashLa (None, 1152, 8)           0         
_________________________________________________________________
caps2_predicted (MyDigitCaps (None, 1152, 10, 16, 1)   1474560   
_________________________________________________________________
routing1 (RoutingByAgreement (None, 1, 10, 16, 1)      11520    

In [ ]:
!du -sh /content/drive/MyDrive/MnistResults/*

27M	/content/drive/MyDrive/MnistResults/best_weights4.hdf5
27M	/content/drive/MyDrive/MnistResults/best_weights5.hdf5
1.0K	/content/drive/MyDrive/MnistResults/caps_basemodelA.h5
27M	/content/drive/MyDrive/MnistResults/caps_basemodelA.tf
27M	/content/drive/MyDrive/MnistResults/caps_best_weightsA.hdf5
27M	/content/drive/MyDrive/MnistResults/caps_latest_weightsA.hdf5
4.0K	/content/drive/MyDrive/MnistResults/caps_mylogsA.csv
27M	/content/drive/MyDrive/MnistResults/caps_pruned_fileA.h5
95K	/content/drive/MyDrive/MnistResults/cnn_best_weights4.hdf5
95K	/content/drive/MyDrive/MnistResults/cnn_latest_weights4.hdf5
1.0K	/content/drive/MyDrive/MnistResults/cnn_mylogs4.csv
367K	/content/drive/MyDrive/MnistResults/cnn_save_basemodel4.tf
11M	/content/drive/MyDrive/MnistResults/compressed1.tflite
21M	/content/drive/MyDrive/MnistResults/dense_best_weights1.hdf5
21M	/content/drive/MyDrive/MnistResults/dense_latest_weights1.hdf5
4.0K	/content/drive/MyDrive/MnistResults/dense_mylogs1.csv
27M	/content/dr

In [ ]:
from tensorflow.keras.models import load_model

model = model_for_export
import numpy as np

for i, w in enumerate(model.get_weights()):
  print("{} -- Total:{}, Zeros: {:.2f}%".format(model.weights[i].name, w.size, np.sum(w == 0) / w.size * 100))
    

conv_layer_1/kernel:0 -- Total:20736, Zeros: 64.99%
conv_layer_1/bias:0 -- Total:256, Zeros: 0.00%
conv_layer_2/kernel:0 -- Total:5308416, Zeros: 64.99%
conv_layer_2/bias:0 -- Total:256, Zeros: 0.00%
caps2_predicted/kernel:0 -- Total:1474560, Zeros: 64.99%
routing1/raw_weights:0 -- Total:11520, Zeros: 64.99%


In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model_for_export)
converter.optimizations = [tf.lite.Optimize.DEFAULT]

converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]

tflite_model = converter.convert()


INFO:tensorflow:Assets written to: /tmp/tmpa11q0z5e/assets


AttributeError: ignored

tyoes of post training quantization
https://www.tensorflow.org/lite/performance/post_training_quantization

In [ ]:
def get_gzipped_model_size(file):
  # Returns size of gzipped model, in bytes.
  import os
  import zipfile

  _, zipped_file = tempfile.mkstemp('.zip')
  with zipfile.ZipFile(zipped_file, 'w', compression=zipfile.ZIP_DEFLATED) as f:
    f.write(file)

  return os.path.getsize(zipped_file)

In [ ]:
#converter = tf.lite.TFLiteConverter.from_keras_model(model_for_export)
tflite_model_file = "/content/drive/MyDrive/MnistResults/caps_compressedA.tflite"
converter = tf.lite.TFLiteConverter.from_keras_model(model_for_export)
converter.optimizations = [tf.lite.Optimize.DEFAULT]

converter.target_spec.supported_ops = [
  tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
]

tflite_model = converter.convert()
with open(tflite_model_file, "wb") as f:
    f.write(tflite_model)



INFO:tensorflow:Assets written to: /tmp/tmp5dojix44/assets


INFO:tensorflow:Assets written to: /tmp/tmp5dojix44/assets


In [ ]:
#print("Size of gzipped baseline Keras model: %.2f bytes" % (get_gzipped_model_size(basemodel_file)))
print("Size of gzipped pruned Keras model: %.2f bytes" % (get_gzipped_model_size(pruned_keras_file)))
print("Size of gzipped pruned TFlite model: %.2f bytes" % (get_gzipped_model_size(tflite_model_file)))

Size of gzipped pruned Keras model: 11882929.00 bytes
Size of gzipped pruned TFlite model: 4808114.00 bytes


In [ ]:
import numpy as np

def evaluate_model(interpreter):
  input_index = interpreter.get_input_details()[0]["index"]
  output_index = interpreter.get_output_details()[0]["index"]

  # Run predictions on ever y image in the "test" dataset.
  prediction_digits = []
  for i, test_image in enumerate(x_test[:10]):
    if i % 1000 == 0:
      print('Evaluated on {n} results so far.'.format(n=i))
    # Pre-processing: add batch dimension and convert to float32 to match with
    # the model's input data format.
    test_image = np.expand_dims(test_image, axis=0).astype(np.float32)
    interpreter.set_tensor(input_index, test_image)

    # Run inference.
    interpreter.invoke()

    # Post-processing: remove batch dimension and find the digit with highest
    # probability.
    output = interpreter.tensor(output_index)
    digit = np.argmax(output()[0])
    prediction_digits.append(digit)

  print('\n')
  # Compare prediction results with ground truth labels to calculate accuracy.
  prediction_digits = np.array(prediction_digits)
  accuracy = (prediction_digits == y_test[:10]).mean()
  return accuracy

In [ ]:
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

test_accuracy = evaluate_model(interpreter)

print('Pruned and quantized TFLite test_accuracy:', test_accuracy)
print('Pruned TF test accuracy:', model_for_pruning_accuracy)

Evaluated on 0 results so far.


ValueError: ignored

In [ ]:
modelB = keras.models.load_model()

In [ ]:
modelB = K.models.load_model('/content/drive/MyDrive/WeightsMnist/save2.tf',\
                         custom_objects=\
                          {\
                              "SquashLayer": SquashLayer,\
                              "SafeNorm": SafeNorm,\
                              "MyDigitCapsLayer": MyDigitCapsLayer,\
                              "RoutingByAgreement": RoutingByAgreement,\
                              "MyAccuracy": MyAccuracy,\
                              "MarginLoss": MarginLoss,\
                          })


TypeError: ignored